### PR TITLE
[ios] Using `UnsafeBufferPointer` to skip data copying and boost performance.

### PR DESCRIPTION
**Pull request by kikiloveswift**
# Optimizing Spine Performance on iOS: Reducing CPU Costs with UnsafeBufferPointer
Recently, I tested the performance of Spine on an iPad and noticed that it has high CPU usage. To dig deeper into the performance costs, I used Xcode Instruments' Time Profiler. After analyzing the data, I discovered that the most CPU-intensive function was `RenderCommand.getVertices()`.

One of the main issues I identified was the use of compactMap, which adds extra overhead during array generation. In contrast, using UnsafeBufferPointer helps to avoid this performance penalty, significantly reducing CPU usage.

After testing various Spine files and animations, I found that switching to UnsafeBufferPointer resulted in a 12% to 20% reduction in CPU usage. This optimization worked well for both animations and skins, improving performance across the board.

Additionally, I found that pre-allocating array capacity with `vertices.reserveCapacity(indices.count)` is also beneficial. This prevents the need for array resizing when new elements are appended, which can also incur high CPU costs.

# Performance Comparison: spine-boy Walk Animation
Here’s a comparison of the performance when using the [spine-boy](https://github.com/EsotericSoftware/spine-runtimes/tree/4.2/spine-ios/Example%20-%20Cocoapods/Spine%20iOS%20Example/spineboy) `walk` animation:

**Use `compactMap` to generate an array(By default):**
image

**Use `UnsafeBufferPointer ` :**
image

### CODE REVIEW

**Comment by badlogic** [comment]
This looks great! I was actually contemplating switching to `UnsafeBufferPointer` but didn't find the time to look into it yet. I'll have to test this a bit before I merge, but it looks good to me!

**Comment by kikiloveswift** [comment]
> This looks great! I was actually contemplating switching to `UnsafeBufferPointer` but didn't find the time to look into it yet. I'll have to test this a bit before I merge, but it looks good to me!

Thanks for taking a review! I’m glad it aligns with what you were considering. Let me know if you encounter anything during testing or if there’s anything I can refine further.

File: spine-cpp/spine-cpp-lite/spine-cpp-lite-codegen.py [hunk 1]
```diff
@@ -305,24 +303,45 @@ def write_c_function_call(self, spine_params):
...
 
       return function_call
-    
-  def write_array_call(self, num_function_name, function_call):
+  
+  def write_array_spine_class(self, num_function_name, function_call):
     array_call = f"let num = Int({num_function_name}({self.spine_object.var_name}))"
     array_call += "\n"
     array_call += inset + inset
     array_call += f"let ptr = {function_call}"
```
Comment (denrase) [inline spine-cpp/spine-cpp-lite/spine-cpp-lite-codegen.py]: If we do the guard let earlier here, we don't need a second one. Would this work?

```swift
array_call += f"guard let ptr = {function_call} else { return [] }"
```

File: spine-ios/Sources/Spine/Spine.Generated.swift [hunk 2]
```diff
@@ -26,8 +26,10 @@ public final class TransformConstraintData: NSObject {
     public var bones: [BoneData] {
```
Comment (denrase) [inline spine-ios/Sources/Spine/Spine.Generated.swift:26]: Do you think it makes sense to simplify the code-gen so we call `compactMap` directly on the buffer, would this work?

```swift
public var bones: [BoneData] {
    let num = Int(spine_skeleton_data_get_num_bones(wrappee))
    guard let ptr = spine_skeleton_data_get_bones(wrappee) else { return [] }
    let buffer = UnsafeBufferPointer(start: ptr, count: num)
    return buffer.compactMap { $0.flatMap { .init($0) } }
}
```

It looks like we are copying the `UnsafeBufferPointer`s content two times. Once with Array(buffer) and once with the `.compactMap`?

**Review by denrase — CHANGES_REQUESTED** [review]
Thank you for doing this, awesome CPU impovement! 🥳  Just some minor questions/suggestions from my side.

@badlogic From an ARC perspective, this pretty much looks the same to me as before. The `.compactMap` did copy the values into an array, as does the `Array(buffer)`. [UnsafeBufferPointer](https://developer.apple.com/documentation/swift/unsafebufferpointer#UnsafeBufferPointer-Semantics) does not own/retain any data if i understood correctly.

File: spine-cpp/spine-cpp-lite/spine-cpp-lite-codegen.py (hunk 1 above)
Comment (kikiloveswift) [reply spine-cpp/spine-cpp-lite/spine-cpp-lite-codegen.py]: Yes, you're right.

File: spine-ios/Sources/Spine/Spine.Generated.swift (hunk 2 above)
Comment (kikiloveswift) [reply spine-ios/Sources/Spine/Spine.Generated.swift:26]: Yes, you're right. Creating an array with `Array(buffer)` indeed involves a memory copy. Thanks for pointing that out!

**Comment by kikiloveswift** [comment]
> Thank you for doing this, awesome CPU impovement! 🥳 Just some minor questions/suggestions from my side.
> 
> @badlogic From an ARC perspective, this pretty much looks the same to me as before. The `.compactMap` did copy the values into an array, as does the `Array(buffer)`. [UnsafeBufferPointer](https://developer.apple.com/documentation/swift/unsafebufferpointer#UnsafeBufferPointer-Semantics) does not own/retain any data if i understood correctly.

Thanks for your review! `UnsafeBufferPointer` doesn’t own or retain the data, and both methods indeed copy the data.

`compactMap:`
- Involves converting each element to `Float` and building a new array, which adds some computational overhead.


`Array(buffer)` and `UnsafeBufferPointer`:
- `UnsafeBufferPointer` provides direct memory access. Using Array(buffer) to build an array directly from the buffer can be more optimized internally, as it pre-determines the data count and allocates the right amount of memory, reducing memory reallocation and copying.

**Comment by badlogic** [comment]
Super! I put it through its paces as well via Instruments and everything works as intended.

Thanks!

**Pull request merged**